pkg/workspace/inference/max_model_len_strategy.go

```diff
@@ -0,0 +1,151 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package inference
+
+import (
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+
+	pkgmodel "github.com/kaito-project/kaito/pkg/model"
+	"github.com/kaito-project/kaito/pkg/sku"
+)
+
+// computeMaxModelLen calculates the optimal max model length for GPU memory efficiency.
+func computeMaxModelLen(preset *pkgmodel.PresetParam, gpu *sku.GPUConfig, numRequiredNodes int) int {
+	// Validate input parameters
+	if preset == nil || gpu == nil || numRequiredNodes <= 0 {
+		return 0
+	}
+	if preset.ModelTokenLimit <= 0 || preset.BytesPerToken <= 0 || gpu.GPUMemGB <= 0 || gpu.GPUCount <= 0 {
+		return 0
+	}
+
+	// Parse model weight size using Kubernetes resource.Quantity
+	weightGiB, ok := parseModelWeight(preset.TotalSafeTensorFileSize)
+	if !ok {
+		return 0
+	}
+
+	// Calculate available GPU memory and adjusted bytes per token
+	availableBytes, adjustedBytesPerToken := calculateMemoryParameters(preset, gpu, numRequiredNodes, weightGiB)
+	if availableBytes <= 0 || adjustedBytesPerToken <= 0 {
+		return 0
+	}
+
+	// Calculate raw token candidate
+	candidate := int(availableBytes / adjustedBytesPerToken)
+	if candidate <= 0 {
+		return 0
+	}
+
+	// Apply constraints and finalize result
+	finalResult := applyConstraintsAndAlignment(candidate, preset.ModelTokenLimit)
+
+	klog.Infof("computeMaxModelLen: final result=%d", finalResult)
+	return finalResult
+}
+
+// parseModelWeight parses the model weight string and returns the weight in GiB.
+// It handles Kubernetes resource.Quantity format strings like "25.63Gi", etc.
+func parseModelWeight(totalSafeTensorFileSize string) (float64, bool) {
+	s := strings.TrimSpace(totalSafeTensorFileSize)
+	if s == "" {
+		return 0, false
+	}
+
+	// Use Kubernetes resource.Quantity to parse the size string
+	quantity, err := resource.ParseQuantity(s)
+	if err != nil {
+		return 0, false
+	}
+
+	bytes := quantity.Value()
+	if bytes <= 0 {
+		return 0, false
+	}
+
+	// Convert bytes to GiB (1 GiB = 2^30 bytes)
+	weightGiB := float64(bytes) / (1 << 30)
+	if weightGiB <= 0 {
+		return 0, false
+	}
+
+	return weightGiB, true
+}
+
+// calculateMemoryParameters computes available GPU memory and adjusted bytes per token.
+// Returns the available memory in bytes and the adjusted bytes per token for the calculation.
+func calculateMemoryParameters(preset *pkgmodel.PresetParam, gpu *sku.GPUConfig, numRequiredNodes int, weightGiB float64) (float64, float64) {
```

Comment:
You can simplify things by viewing it as a whole.

avaialble_mem_for_kvcache = (per_gpu_memory - non_torch_overhead) * num_gpu - activation - model_weights.